cmd/crank/beta/trace/internal/resource/xrm/loader.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2023 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package xrm
+
+import (
+	"context"
+	"sort"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/crossplane/crossplane/cmd/crank/beta/trace/internal/resource"
+)
+
+// channelCapacity is the buffer size of the processing channel, should be a high value
+// so that there is no blocking. Correctness of processing does not depend on the channel capacity.
+var channelCapacity = 1000 //nolint:gochecknoglobals // we treat this as constant only overrideable for tests.
+
+// workItem maintains the relationship of a resource to be loaded with its parent
+// such that the resource that is loaded can be added as a child.
+type workItem struct {
+	parent *resource.Resource
+	child  v1.ObjectReference
+}
+
+// resourceLoader is a delegate that loads resources and returns child resource refs.
+type resourceLoader interface {
+	loadResource(ctx context.Context, ref *v1.ObjectReference) *resource.Resource
+	getResourceChildrenRefs(_ context.Context, r *resource.Resource) []v1.ObjectReference
```

Comment:
I kept it in the interface although the implementation does not need it to be symmetric with the other method. Also it should be `ctx` here and the `_` identifier in the implementation since it doesn't use it. I can remove this as well.

Reply:
fixed in df75bcaaae93e7a956d44f0c62fe3a7982b70f19